### DISCUSSION TITLE
Map a nested field using fields from parent

### DISCUSSION THREAD

**Discussion by JanecekPetr** [mapstruct/mapstruct#3442] [Q&A] [answered]
Basically, this: https://stackoverflow.com/questions/69423911

Also I saw https://stackoverflow.com/questions/53878367 and https://stackoverflow.com/questions/74522692 and https://stackoverflow.com/questions/50997281

I have a parent and child class structure
```java
record Parent(
    int number,
    List<Child> children
) {}

record Child(
    String name
) {}
```

and DTOs, however here the child contains something from the Parent
```java
record ParentDto(
    int number,
    List<ChildDto> children
) {}

record ChildDto(
    String name,
    int number     // <-- this is the field that gives seems impossible
) {}
```

It currently seems impossible to map a `Parent` to `ParentDto` using MapStruct in such a way that ChildDto would contain the `number` field from their parent. The workarounds posted in the above mentioned SO questions only work with mutable classes by calling a setter in an @AfterMapping method.

Is this really impossible with an immutable structure?

I'd be very happy with providing a manual mapping method for either Parent or Child, however in the real-world case the structure is complex, contains many classes and many levels. It seems impossible to compose the automatic and manual code in a way that MapStruct could help me in any way, even with partial mappings =/.

**Comment by filiphr** [comment, marked as answer]
This is a tricky. It can be achieve using `@Context` I believe.

e.g.

```java
@Mapper
public abstract class ParentMapper {

    public ParentDto map(Parent parent) {
        return mapWithParentContext(parent, parent);
    }

    @Named("withContext")
    protected abstract ParentDto mapWithParentContext(Parent parent, @Context Parent parentContext);

    protected ChildDto mapChildWithParentContext(Child child, @Context Parent parent) {
        return map(child, parent);
    }

    @Named("child")
    @Mapping(target = "number", source = "parent.number")
    protected abstract ChildDto map(Child child, Parent parent);
}
```

**Reply by JanecekPetr** [reply]
Oh that's a good trick! I tried some Context shenanigans, but this never came up. I'll consider it on the next iteration, thanks!

In the end I went with something along the lines of:

```
@Mapper(uses = { ChildMapper.class })
public interface ParentMapper {
    @Mapping(target = "children", source = "parentDto")
    Parent fromDto(ParentDto parentDto);
}

@Mapper
public interface ChildMapper {
    @Mapping(target = "number", source = "parentDto.number")
    Child fromDto(ChildDto childDto, ParentDto parentDto);

    default List<Child> fromParentDto(ParentDto parentDto) {
        return parentDto.children()
                .stream()
                .map(childDto -> fromDto(childDto, parentDto))
                .toList();
    }
}
```
